monty-31/src/dft/backward.rs

```diff
@@ -8,33 +8,206 @@ extern crate alloc;
+        // combined into one, was not only not faster, but was
+        // actually a bit slower.
+
+        // Radix 2
+        backward_iterative_packed_radix_2(input);
+
+        // Radix 4
+        let roots4 = [MP::INV_ROOTS_8.as_ref()[0], MP::INV_ROOTS_8.as_ref()[2]];
+        if <Self as Field>::Packing::WIDTH >= 4 {
+            backward_iterative_packed::<2, _>(input, &roots4);
+        } else {
+            Self::backward_iterative_layer(input, &roots4, 2);
+        }
+
+        // Radix 8
+        if <Self as Field>::Packing::WIDTH >= 8 {
+            backward_iterative_packed::<4, _>(input, MP::INV_ROOTS_8.as_ref());
+        } else {
+            Self::backward_iterative_layer(input, MP::INV_ROOTS_8.as_ref(), 4);
+        }
+
+        // Radix 16
+        if <Self as Field>::Packing::WIDTH >= 16 {
+            backward_iterative_packed::<8, _>(input, MP::INV_ROOTS_16.as_ref());
+        } else {
+            Self::backward_iterative_layer(input, MP::INV_ROOTS_16.as_ref(), 8);
+        }
+    }
+
+    fn backward_iterative(input: &mut [Self], root_table: &[Vec<Self>]) {
+        let n = input.len();
+        let lg_n = log2_strict_usize(n);
+
+        // Needed to avoid overlap with specialisation at the other end of the loop
+        assert!(lg_n >= 6);
+
+        let packing_width = <Self as Field>::Packing::WIDTH;
+        assert!(n >= 2 * packing_width);
+
+        let packed_input = <Self as Field>::Packing::pack_slice_mut(input);
+
+        // Start loop after doing radix 16 separately. This value is determined by the largest
+        // packing width we will encounter, which is 16 at the moment for AVX512. Specifically
+        // it is log_2(max{possible packing widths}) = lg(16) = 4.
+        const FIRST_LOOP_LAYER: usize = 4;
+
+        // How many layers have we specialised after the main loop
+        const NUM_SPECIALISATIONS: usize = 2;
+
+        // Needed to avoid overlap of the 2 specialisations at the start
+        // with the radix-16 specialisation at the end of the loop
+        assert!(lg_n >= FIRST_LOOP_LAYER + NUM_SPECIALISATIONS);
+
+        Self::backward_iterative_packed_radix_16(packed_input);
+
+        for lg_m in FIRST_LOOP_LAYER..(lg_n - NUM_SPECIALISATIONS) {
+            let s = lg_n - lg_m - 1;
+            let m = 1 << lg_m;
+
+            let roots = &root_table[s];
+            debug_assert_eq!(roots.len(), m);
+
+            Self::backward_iterative_layer(packed_input, roots, m);
+        }
+        // Specialise the last few iterations; improves performance a little.
+        backward_iterative_layer_1(packed_input, &root_table[1]); // lg_m == lg_n - 2, s == 1
+        backward_iterative_layer_0(packed_input, &root_table[0]); // lg_m == lg_n - 1, s == 0
+    }
 
-impl<MP: MontyParameters + TwoAdicData> MontyField31<MP> {
     #[inline(always)]
     fn backward_butterfly(x: Self, y: Self, w: Self) -> (Self, Self) {
         let t = y * w;
         (x + t, x - t)
     }
```

Comment:
This function is a duplicate of `backward_butterfly` defined at the top of this file. Should delete one of them.